job-server/src/spark.jobserver/JobCache.scala

```diff
@@ -22,28 +26,51 @@ class JobCacheImpl(maxEntries: Int,
                    loader: ContextURLClassLoader) extends JobCache {
   import scala.concurrent.duration._
 
-  private val cache = new LRUCache[(String, DateTime, String), JobJarInfo](maxEntries)
+  private val cache = new LRUCache[(String, DateTime, String), BinaryJobInfo](maxEntries)
   private val logger = LoggerFactory.getLogger(getClass)
   implicit val daoAskTimeout: Timeout = Timeout(3 seconds)
 
   /**
    * Retrieves the given SparkJob class from the cache if it's there, otherwise use the DAO to retrieve it.
-   * @param appName the appName under which the jar was uploaded
-   * @param uploadTime the upload time for the version of the jar wanted
+   * @param appName the appName under which the binary was uploaded
+   * @param uploadTime the upload time for the version of the binary wanted
    * @param classPath the fully qualified name of the class/object to load
    */
   def getSparkJob(appName: String, uploadTime: DateTime, classPath: String): JobJarInfo = {
-    cache.get((appName, uploadTime, classPath), {
-      import akka.pattern.ask
-      import scala.concurrent.Await
-
-      val jarPathReq = (dao ? JobDAOActor.GetJarPath(appName, uploadTime)).mapTo[JobDAOActor.JarPath]
-      val jarPath = Await.result(jarPathReq, daoAskTimeout.duration).jarPath
-      val jarFilePath = new java.io.File(jarPath).getAbsolutePath()
-      sparkContext.addJar(jarFilePath) // Adds jar for remote executors
-      loader.addURL(new URL("file:" + jarFilePath)) // Now jar added for local loader
-      val constructor = JarUtils.loadClassOrObject[spark.jobserver.api.SparkJobBase](classPath, loader)
-      JobJarInfo(constructor, classPath, jarFilePath)
-    })
+    cache.get((appName, uploadTime, classPath)) match {
```

Comment:
Huh.  I guess we removed spray-caching at some point.  Let's add that
back.  Spray caching is exactly what you describe - Futures and Akka-based,
with atomic operations.

I'm also OK using JDK8's ConcurrentHashMap with the computeIfAbsent
methods, which are similar.  We'd have to be OK with being a JDK8-only
codebase....

On Fri, Aug 12, 2016 at 2:10 AM, mattinbits <notifications@github.com>
wrote:

> In job-server/src/spark.jobserver/JobCache.scala
> <https://github.com/spark-jobserver/spark-jobserver/pull/553#discussion_r74561596>
> :
>
> >     * @param classPath the fully qualified name of the class/object to load
> >     */
> >    def getSparkJob(appName: String, uploadTime: DateTime, classPath: String): JobJarInfo = {
> > -    cache.get((appName, uploadTime, classPath), {
> > -      import akka.pattern.ask
> > -      import scala.concurrent.Await
> > -
> > -      val jarPathReq = (dao ? JobDAOActor.GetJarPath(appName, uploadTime)).mapTo[JobDAOActor.JarPath]
> > -      val jarPath = Await.result(jarPathReq, daoAskTimeout.duration).jarPath
> > -      val jarFilePath = new java.io.File(jarPath).getAbsolutePath()
> > -      sparkContext.addJar(jarFilePath) // Adds jar for remote executors
> > -      loader.addURL(new URL("file:" + jarFilePath)) // Now jar added for local loader
> > -      val constructor = JarUtils.loadClassOrObject[spark.jobserver.api.SparkJobBase](classPath, loader)
> > -      JobJarInfo(constructor, classPath, jarFilePath)
> > -    })
> > +    cache.get((appName, uploadTime, classPath)) match {
>
> I am happy to revert to the original style, but I agree with @hntd187
> <https://github.com/hntd187> the current cache implementation is not
> futures based and the get(...) in the original implementation does not
> appear atomic to me. There's no synchronisation and
> java.util.LinkedHashMap is not a collection designed for concurrency.
>
> Since we are using Akka, I think the appropriate way to get an atomic
> futures based cache would be to encapsulate the JobCache in an Actor (for
> atomicity) and use the ask pattern (for futures). (not in this PR though
> 😄 )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/553/files/b4ea3b476442b188582c22f86fe5e18e504de3b4#r74561596>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABA321EkQkhl1RkfWL-0BOQPRzdGyINzks5qfDhpgaJpZM4JZq4L>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison